File: 1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.c
#define MAX_ROWS 300
#define MAX_COLS 300

// Hash function for string
unsigned long hash(char* str, int len) {
    unsigned long hash = 5381;
    for (int i = 0; i < len; i++) {
        hash = ((hash << 5) + hash) + str[i];
    }
    return hash;
}

// Hash table structure
typedef struct {
    char key[MAX_COLS + 1];
    int count;
} HashEntry;

int maxEqualRowsAfterFlips(int** matrix, int matrixSize, int* matrixColSize) {
    HashEntry hashTable[MAX_ROWS * 2];
    int hashTableSize = 0;
    int maxEqualRows = 0;
    
    for (int i = 0; i < matrixSize; i++) {
        char original[MAX_COLS + 1] = {0};
        char flipped[MAX_COLS + 1] = {0};
        
        // Create original and flipped patterns
        for (int j = 0; j < matrixColSize[0]; j++) {
            original[j] = matrix[i][j] + '0';
            flipped[j] = (1 - matrix[i][j]) + '0';
        }
        
        // Choose lexicographically smaller pattern
        char* key = strcmp(original, flipped) <= 0 ? original : flipped;
        
        // Find or add to hash table
        int found = 0;
        for (int j = 0; j < hashTableSize; j++) {
            if (strcmp(hashTable[j].key, key) == 0) {
                hashTable[j].count++;
                maxEqualRows = fmax(maxEqualRows, hashTable[j].count);
                found = 1;
                break;
            }
        }
        
        // If not found, add new entry
        if (!found) {
            strcpy(hashTable[hashTableSize].key, key);
            hashTable[hashTableSize].count = 1;
            maxEqualRows = fmax(maxEqualRows, 1);
            hashTableSize++;
        }
    }
    
    return maxEqualRows;
}

File: 1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.java
class Solution {
    public int maxEqualRowsAfterFlips(int[][] matrix) {
        Map<String, Integer> patternCount = new HashMap<>();
        int maxEqualRows = 0;
        
        for (int[] row : matrix) {
            // Create two possible patterns: original and flipped
            StringBuilder original = new StringBuilder();
            StringBuilder flipped = new StringBuilder();
            
            for (int cell : row) {
                original.append(cell);
                flipped.append(1 - cell);
            }
            
            // Generate pattern string
            String pattern = original.toString();
            String flippedPattern = flipped.toString();
            
            // Choose lexicographically smaller pattern
            String key = pattern.compareTo(flippedPattern) <= 0 ? pattern : flippedPattern;
            
            // Count pattern occurrences
            patternCount.put(key, patternCount.getOrDefault(key, 0) + 1);
            maxEqualRows = Math.max(maxEqualRows, patternCount.get(key));
        }
        
        return maxEqualRows;
    }
}